File: python/day16.py
with open("../inputs/day16.txt") as file:
    grid = file.read().splitlines()


def calc_energized(grid: list[str], start: tuple[int, int, int, int]) -> int:
    # (row, col, movement row, movement col)
    queue = [start]
    seen: set[tuple[int, int, int, int]] = set()

    while queue:
        row, col, drow, dcol = queue.pop(0)
        row += drow
        col += dcol

        if row < 0 or row >= len(grid) or col < 0 or col >= len(grid[0]):
            continue

        new_pos = grid[row][col]

        if (
            new_pos == "."
            or (new_pos == "-" and dcol != 0)
            or (new_pos == "|" and drow != 0)
        ):
            queue.append((row, col, drow, dcol))
            seen.add((row, col, drow, dcol))

        elif new_pos == "\\":
            drow, dcol = dcol, drow
            if (row, col, drow, dcol) not in seen:
                queue.append((row, col, drow, dcol))
                seen.add((row, col, drow, dcol))

        elif new_pos == "/":
            drow, dcol = -dcol, -drow
            if (row, col, drow, dcol) not in seen:
                queue.append((row, col, drow, dcol))
                seen.add((row, col, drow, dcol))

        else:
            for dr, dc in [(1, 0), (-1, 0)] if new_pos == "|" else [(0, 1), (0, -1)]:
                if (row, col, dr, dc) not in seen:
                    queue.append((row, col, dr, dc))
                    seen.add((row, col, dr, dc))

    visited = {(row, col) for (row, col, _, _) in seen}

    return len(visited)


# row -1 because its like you start outside the grid
solution1 = calc_energized(grid, (0, -1, 0, 1))

solution2 = 0
for row in range(len(grid)):
    solution2 = max(
        solution2,
        calc_energized(grid, (row, -1, 0, 1)),
        calc_energized(grid, (row, len(grid), 0, -1)),
    )

for col in range(len(grid[0])):
    solution2 = max(
        solution2,
        calc_energized(grid, (-1, col, 1, 0)),
        calc_energized(grid, (len(grid), col, -1, 0)),
    )

File: rust/2023/src/days/day16.rs
use std::collections::HashSet;

use crate::{
    advent_tests,
    advent_utils::{AdventSolution, GenericResult},
};

pub struct Solution;

fn calc_energized(input: &[&[u8]], start: (isize, isize, isize, isize)) -> usize {
    let mut queue = vec![start];
    let mut seen = HashSet::new();

    while let Some((mut row, mut col, drow, dcol)) = queue.pop() {
        row += drow;
        col += dcol;

        if row < 0 || row as usize >= input.len() || col < 0 || col as usize >= input[0].len() {
            continue;
        }

        let new_pos = input[row as usize][col as usize];

        match new_pos {
            b'.' => {
                queue.push((row, col, drow, dcol));
                seen.insert((row, col, drow, dcol));
            }
            b'\\' if !seen.contains(&(row, col, dcol, drow)) => {
                queue.push((row, col, dcol, drow));
                seen.insert((row, col, dcol, drow));
            }
            b'/' if !seen.contains(&(row, col, -dcol, -drow)) => {
                queue.push((row, col, -dcol, -drow));
                seen.insert((row, col, -dcol, -drow));
            }
            b'|' if drow != 0 => {
                queue.push((row, col, drow, dcol));
                seen.insert((row, col, drow, dcol));
            }
            b'|' => {
                if !seen.contains(&(row, col, 1, 0)) {
                    queue.push((row, col, 1, 0));
                    seen.insert((row, col, 1, 0));
                }
                if !seen.contains(&(row, col, -1, 0)) {
                    queue.push((row, col, -1, 0));
                    seen.insert((row, col, -1, 0));
                }
            }
            b'-' if dcol != 0 => {
                queue.push((row, col, drow, dcol));
                seen.insert((row, col, drow, dcol));
            }
            b'-' => {
                if !seen.contains(&(row, col, 0, 1)) {
                    queue.push((row, col, 0, 1));
                    seen.insert((row, col, 0, 1));
                }
                if !seen.contains(&(row, col, 0, -1)) {
                    queue.push((row, col, 0, -1));
                    seen.insert((row, col, 0, -1));
                }
            }
            _ => continue,
        }
    }
    seen.iter()
        .map(|(row, col, _, _)| (row, col))
        .collect::<HashSet<_>>()
        .len()
}

impl AdventSolution for Solution {
    fn part1(input: Vec<String>) -> GenericResult<impl std::fmt::Display> {
        let new_input = input.iter().map(|s| s.as_bytes()).collect::<Vec<_>>();
        Ok(calc_energized(&new_input, (0, -1, 0, 1)))
    }

    fn part2(input: Vec<String>) -> GenericResult<impl std::fmt::Display> {
        let new_input = input.iter().map(|s| s.as_bytes()).collect::<Vec<_>>();
        let result = (0..new_input.len()).fold(0, |acc, n| {
            acc.max(calc_energized(&new_input, (n as isize, -1, 0, 1)))
                .max(calc_energized(
                    &new_input,
                    (n as isize, new_input.len() as isize, 0, -1),
                ))
        });
        Ok((0..new_input[0].len()).fold(result, |acc, n| {
            acc.max(calc_energized(&new_input, (-1, n as isize, 1, 0)))
                .max(calc_energized(
                    &new_input,
                    (new_input.len() as isize, n as isize, -1, 0),
                ))
        }))
    }
}

advent_tests!(
    part 1 => (
        "../../tests/2023/day16.txt" => 46
    ),
    part 2 => (
        "../../tests/2023/day16.txt" => 51
    )
);
